File: 03-coding-patterns/examples/15-longest-common-subseq/solution.cpp
// Longest Common Subsequence — C++20/23. 2D DP with rolling rows.
#include <algorithm>
#include <iostream>
#include <string>
#include <string_view>
#include <vector>
using namespace std;

int longest_common_subsequence(string_view a, string_view b) {
    if (a.size() < b.size()) swap(a, b);
    int n = static_cast<int>(a.size()), m = static_cast<int>(b.size());
    vector<int> prev(m + 1, 0), curr(m + 1, 0);
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            if (a[i - 1] == b[j - 1]) curr[j] = prev[j - 1] + 1;
            else curr[j] = max(prev[j], curr[j - 1]);
        }
        swap(prev, curr);
    }
    return prev[m];
}

int main() {
    auto check = [](bool ok, const char* label) {
        if (!ok) { cerr << "FAIL: " << label << "\n"; exit(1); }
    };

    check(longest_common_subsequence("abcde", "ace") == 3, "case-1");
    check(longest_common_subsequence("abc", "abc") == 3, "identical");
    check(longest_common_subsequence("abc", "def") == 0, "disjoint");
    check(longest_common_subsequence("", "abc") == 0, "empty-1");
    check(longest_common_subsequence("abc", "") == 0, "empty-2");
    check(longest_common_subsequence("aggtab", "gxtxayb") == 4, "case-hard");

    cout << "PASS\n";
}

File: 03-coding-patterns/examples/15-longest-common-subseq/solution.ts
// Longest Common Subsequence — TypeScript. 2D DP with rolling rows.
function longestCommonSubsequence(text1: string, text2: string): number {
    if (text1.length < text2.length) [text1, text2] = [text2, text1];
    const n = text1.length, m = text2.length;
    let prev = new Array<number>(m + 1).fill(0);
    let curr = new Array<number>(m + 1).fill(0);
    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= m; j++) {
            if (text1[i - 1] === text2[j - 1]) {
                curr[j] = prev[j - 1]! + 1;
            } else {
                curr[j] = Math.max(prev[j]!, curr[j - 1]!);
            }
        }
        [prev, curr] = [curr, prev];
    }
    return prev[m]!;
}

function check(ok: boolean, label: string): void {
    if (!ok) { console.error(`FAIL: ${label}`); process.exit(1); }
}

check(longestCommonSubsequence("abcde", "ace") === 3, "case-1");
check(longestCommonSubsequence("abc", "abc") === 3, "identical");
check(longestCommonSubsequence("abc", "def") === 0, "disjoint");
check(longestCommonSubsequence("", "abc") === 0, "empty-1");
check(longestCommonSubsequence("abc", "") === 0, "empty-2");
check(longestCommonSubsequence("aggtab", "gxtxayb") === 4, "case-hard");

console.log("PASS");
